### PR TITLE
Fix/directory permissions

### PR DESCRIPTION
**Pull request by 149alexniu149**
Previously, permissions issues with directories of tracked files would not be reported and instead display the following message while the agent was running:

"Tailer Progress: Tailer has parsed 0 records (0 bytes), transformed 0 records, skipped 0 records, and has successfully sent 0 records to destination."

This pull request adds warnings for each directory with permissions issues at agent startup. These warnings look like this:

"[WARN] Permission Denied: Agent user unable to read (directory-name)"

This should allow for easier troubleshooting of issues like #58 

I've also changed the order in which the log file location is chosen, prioritizing the config file over the default options. This should allow for easier customization of the agent

### CODE REVIEW

File: src/com/amazon/kinesis/streaming/agent/Agent.java [hunk 1]
```diff
@@ -52,8 +52,10 @@ public static void main(String[] args) throws Exception {
         AgentOptions opts = AgentOptions.parse(args);
         String configFile = opts.getConfigFile();
         AgentConfiguration config = tryReadConfigurationFile(Paths.get(opts.getConfigFile()));
-        Path logFile = opts.getLogFile() != null ? Paths.get(opts.getLogFile()) : (config != null ? config.logFile() : null);
-        String logLevel = opts.getLogLevel() != null ? opts.getLogLevel() : (config != null ? config.logLevel() : null);
+
+        //Path logFile = opts.getLogFile() != null ? Paths.get(opts.getLogFile()) : (config != null ? config.logFile() : null);
+        Path logFile = config != null ? config.logFile() : (opts.getLogFile() != null ? Paths.get(opts.getLogFile()) : null);
```
Comment (chaochenq) [inline src/com/amazon/kinesis/streaming/agent/Agent.java]: Why is this change needed?

File: src/com/amazon/kinesis/streaming/agent/processing/processors/CSVToJSONDataConverter.java [hunk 2]
```diff
@@ -44,14 +46,17 @@
  *
  */
 public class CSVToJSONDataConverter implements IDataConverter {
-    
+
+    protected final Logger logger;
```
Comment (chaochenq) [inline src/com/amazon/kinesis/streaming/agent/processing/processors/CSVToJSONDataConverter.java]: This is unnecessary

File: src/com/amazon/kinesis/streaming/agent/processing/processors/CSVToJSONDataConverter.java [hunk 3]
```diff
@@ -44,14 +46,17 @@
...
     private static String FIELDS_KEY = "customFieldNames";
     private static String DELIMITER_KEY = "delimiter";
     private final List<String> fieldNames;
     private final String delimiter;
     private final IJSONPrinter jsonProducer;
+
     
     public CSVToJSONDataConverter(Configuration config) {
+        this.logger = Logging.getLogger(getClass());
```
Comment (chaochenq) [inline src/com/amazon/kinesis/streaming/agent/processing/processors/CSVToJSONDataConverter.java]: unnecessary

File: src/com/amazon/kinesis/streaming/agent/processing/processors/LogToJSONDataConverter.java [hunk 4]
```diff
@@ -26,28 +26,32 @@
...
  * }
- * 
+ *
  * @author chaocheq
  *
  */
 public class LogToJSONDataConverter implements IDataConverter {
-    
+
+    protected final Logger logger;
```
Comment (chaochenq) [inline src/com/amazon/kinesis/streaming/agent/processing/processors/LogToJSONDataConverter.java]: unnecessary

File: src/com/amazon/kinesis/streaming/agent/processing/processors/LogToJSONDataConverter.java [hunk 5]
```diff
@@ -26,28 +26,32 @@
...
     private List<String> fields;
     private ILogParser logParser;
     private IJSONPrinter jsonProducer;
-    
+
     public LogToJSONDataConverter(Configuration config) {
+        this.logger = Logging.getLogger(getClass());
```
Comment (chaochenq) [inline src/com/amazon/kinesis/streaming/agent/processing/processors/LogToJSONDataConverter.java]: unnecessary

File: src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java [hunk 6]
```diff
@@ -114,6 +115,13 @@ public String toString() {
     private void validateDirectory(Path dir) {
         Preconditions.checkArgument(dir != null, "Directory component is empty!");
         // TODO: validate that the directory component has no glob characters
+
+        //check the permissions of every level in the directory path
+        for(Path level : dir.toAbsolutePath()){
+            if(!Files.isReadable(level)){
+                Logging.getLogger(getClass()).warn("Permission Denied: Agent user unable to read " + dir.toString());
```
Comment (chaochenq) [inline src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java]: Probably just log to the error level?

File: src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java [hunk 7]
```diff
@@ -114,6 +115,13 @@ public String toString() {
     private void validateDirectory(Path dir) {
         Preconditions.checkArgument(dir != null, "Directory component is empty!");
         // TODO: validate that the directory component has no glob characters
+
+        //check the permissions of every level in the directory path
+        for(Path level : dir.toAbsolutePath()){
```
Comment (chaochenq) [inline src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java]: I don't think dir.toAbsolutePath() will give you a list of levels, so no point of iteration here.

File: src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java [hunk 8]
```diff
@@ -114,6 +115,13 @@ public String toString() {
...
+            if(!Files.isReadable(level)){
```
Comment (chaochenq) [inline src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java]: what are you trying to check here? do you want to check if the files are permitted to read by agent (the file grants r permission) or if the directory can be listed by the agent (the directory grants x permission)?

File: src/com/amazon/kinesis/streaming/agent/Agent.java (hunk 1 above)
Comment (149alexniu149) [inline src/com/amazon/kinesis/streaming/agent/Agent.java]: This specifies that configs should be loaded from the config file first rather than from the options, it's consistent with the same fix that we made in line 58

File: src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java (hunk 7 above)
Comment (149alexniu149) [inline src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java]: My understanding is that .toAbsolutePath() will return a path object. 

Since the path class extends iterable, this loop should allow us to check each directory individually as per this documentation (See the very end: https://docs.oracle.com/javase/tutorial/essential/io/pathOps.html). Is this correct?

File: src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java (hunk 8 above)
Comment (149alexniu149) [inline src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java]: Since there is a requirement that every directory up to the file needs to be readable by the agent user, this checks whether that is the case. If it isn't it throws a warning.

File: src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java (hunk 6 above)
Comment (149alexniu149) [inline src/com/amazon/kinesis/streaming/agent/tailing/SourceFile.java]: Warnings won't prevent the agent from stopping completely so if there are permissions issues on multiple directories, every directory with an issue will appear in the logs rather than an error on the first restricted directory

File: src/com/amazon/kinesis/streaming/agent/processing/processors/LogToJSONDataConverter.java (hunk 4 above)
Comment (149alexniu149) [inline src/com/amazon/kinesis/streaming/agent/processing/processors/LogToJSONDataConverter.java]: Removed

File: src/com/amazon/kinesis/streaming/agent/processing/processors/LogToJSONDataConverter.java (hunk 5 above)
Comment (149alexniu149) [inline src/com/amazon/kinesis/streaming/agent/processing/processors/LogToJSONDataConverter.java]: Removed

File: src/com/amazon/kinesis/streaming/agent/processing/processors/CSVToJSONDataConverter.java (hunk 3 above)
Comment (149alexniu149) [inline src/com/amazon/kinesis/streaming/agent/processing/processors/CSVToJSONDataConverter.java]: Removed

File: src/com/amazon/kinesis/streaming/agent/processing/processors/CSVToJSONDataConverter.java (hunk 2 above)
Comment (149alexniu149) [inline src/com/amazon/kinesis/streaming/agent/processing/processors/CSVToJSONDataConverter.java]: Removed
